### PR TITLE
Implement parallel cuda::std::lexicographical_compare

### LINKED ISSUE #7752: [FEA]: Implement CUDA backend for parallel `cuda::std::lexicographical_compare`
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

libcu++

### Is your feature request related to a problem? Please describe.

We want to implement a CUDA backend for the following algorithms:

* `std::lexicoraphical_compare` see https://en.cppreference.com/w/cpp/algorithm/lexicographical_compare.html

### Describe the solution you'd like

Hopefully similar to mismatch

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edenfunf**
Resolves #7752. Implements the CUDA backend for `cuda::std::lexicographical_compare`.

## Design

Following the suggestion on #7752: a `cuda::zip_transform_iterator` produces an
`int8_t` per pair (`-1` LESS, `+1` GREATER, `0` EQUIV), `cub::DeviceFind::FindIf`
finds the first non-zero index with early termination, then a 1-element
`cub::DeviceReduce::TransformReduce` (with `cuda::std::identity{}` + `plus<int8_t>{}`
over `int8_t{0}`) reads the state at that position back to host. Final answer is
`state < 0`.

Spec complexity bound (≤ `2 * min(N1, N2)` `comp` calls) is met: Pass 1 short-circuits
at `2k` worst-case for early-termination index `k`; Pass 2 adds 1; total ≤ `2 * min - 1`.

I prototyped four value-recovery variants locally (1-element FindIf, 1-element
TransformReduce, custom 1-thread kernel, two-FindIf baseline). TransformReduce-identity
won consistently by 5–30% in p50 because cub's 1-element reduce path has lower launch
overhead than 1-element FindIf in this regime. Happy to switch shape if you'd prefer
something different.

## Performance (RTX 5070 / sm_120, CUDA 12.9)

Realistic workload — two independent random sequences (≈always diverges at k=0):

| N    | T   | this PR | full-scan reduce baseline |
|----:|----:|--------:|--------------------------:|
| 16M  | int32 | 103 µs |   274 µs |
| 64M  | int32 | 100 µs |   930 µs |
| 64M  | int64 | 110 µs |  1755 µs |

Worst case for the early-termination design (late divergence and equal-all at
scale): within 5–10% of the no-early-termination baseline.

## Tests

- `pstl_lexicographical_compare.cu` (default `less`) — 480 assertions
- `pstl_lexicographical_compare_comp.cu` (custom `Compare`) — 192 assertions

Both cover the `all_types` matrix × 4 SECTIONs (default stream / provided stream /
provided memory_resource / both), plus edge cases: empty vs empty, empty/non-empty
mixed, single-element ranges (equal/less/greater — exercises the `n=1` path through
both passes), prefix in either direction, divergence at `k=0/N/2/N-1`,
`random_access_iterator` wrapper, and `greater<>`/`less<>` duality
(`lex_cmp(s2, s1, greater) == lex_cmp(s1, s2, less)`).

`bench/lexicographical_compare/basic.cu` mirrors `is_heap_until/basic.cu`'s shape
with `2 * violation_point` memory accounting and `ViolationAt ∈ {1.0, 0.5, 0.01}`.

### CODE REVIEW

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+struct __lex_state_fn
+{
+  _Compare __comp_;
+
+  _CCCL_API explicit constexpr __lex_state_fn(_Compare __comp)
+      : __comp_(::cuda::std::move(__comp))
+  {}
+
+  template <class _Tp, class _Up>
+  [[nodiscard]] _CCCL_DEVICE_API constexpr int8_t operator()(const _Tp& __a, const _Up& __b) const
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: @fbusato I believe using plain `int` might be more efficient because we do not need to unpack the `int8_t` thoughts?

**Comment by edenfunf** [comment]
Tested locally on RTX 5070 / sm_120, CUDA 12.9, with both variants
implemented identically except for the state type. Per-call wall-clock
p50 (5 warmup + 50 reps via cudaEvent pairs); ratio = `int` / `int8_t`
so values < 1.0 mean `int` wins.

| N    | T     | workload      | int8_t p50 | int p50 | ratio |
|----:|----:|---------------|-----------:|--------:|------:|
| 64M  | int32 | diverge@k=10  | 101.3 µs   |  79.6 µs | 0.785 |
| 16M  | int64 | diverge@k=10  | 120.6 µs   |  78.3 µs | 0.649 |
| 64M  | int64 | diverge@k=10  | 144.6 µs   | 117.3 µs | 0.811 |
| 16M  | int32 | diverge@k=N/2 | 242.4 µs   | 225.0 µs | 0.928 |
| 64M  | int32 | rand-all-diff |  77.4 µs   |  78.2 µs | 1.010 |
| 64M  | int32 | equal-all     | 916.7 µs   | 928.4 µs | 1.013 |

Honest disclosure: small-N + equal-all (where Pass 2 is skipped because
`k == n`) shows `int8_t` slightly faster by ~10 µs absolute (e.g., N=1M
i32 equal-all: int8_t 44 µs vs int 55 µs). That regime is launch-bound
either way, and the gain on the hot path (early-divergence + large N,
where Pass 2's 1-element TransformReduce dominates) is the meaningful
trade.

Pushed the swap as 15c236d6e3.

**Comment by edenfunf** [comment]
Hi @miscco — gentle ping when you have a moment, I think this is ready for another look.

The int8_t → int suggestion is done (15c236d6), with the before/after numbers in the table above. I also re-checked locally on a 5070 (CUDA 12.9): both tests pass, plus a 50M cross-check against std::lexicographical_compare across int/int64/uint8/float/double and the edge cases, and memcheck came back clean.

Happy to make any changes — please let me know if anything's still blocking. Thanks!

**Comment by miscco** [comment]
/ok to test [15c236d](https://github.com/NVIDIA/cccl/pull/8735/commits/15c236d6e320e688a48e3e05fe20ac0e2eff0eeb)

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  _CCCL_API explicit constexpr __lex_state_fn(_Compare __comp)
+      : __comp_(::cuda::std::move(__comp))
+  {}
+
+  template <class _Tp, class _Up>
+  [[nodiscard]] _CCCL_DEVICE_API constexpr int operator()(const _Tp& __a, const _Up& __b) const
+  {
+    if (__comp_(__a, __b))
+    {
+      return -1;
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: Important: We avoid magical numbers as much as possible. Please create an enum class with members `__less`, `__equal`, `__greater`

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  _Compare __comp = {})
+{
+  static_assert(indirect_strict_weak_order<_Compare, _InputIter1, _InputIter2>,
+                "cuda::std::lexicographical_compare: Compare must satisfy "
+                "indirect_strict_weak_order<Compare, InputIter1, InputIter2>");
+
+  [[maybe_unused]] auto __find_dispatch =
+    ::cuda::std::execution::__pstl_select_dispatch<::cuda::std::execution::__pstl_algorithm::__find_if, _Policy>();
+  [[maybe_unused]] auto __reduce_dispatch =
+    ::cuda::std::execution::__pstl_select_dispatch<::cuda::std::execution::__pstl_algorithm::__transform_reduce,
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: The current approach will allocate twice, because we need temporary storage for find_if and also reduce. 

However, we can coalesce both allocations into one because we know how much storage we need. We can also reuse the temp storage allocation for both reduce and find and only have one additional int sized one for the return value.

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    const auto __n  = ::cuda::std::min(__n1, __n2);
+
+    // Lazy per-element state (-1 / 0 / +1) over the two input ranges, capped at
+    // the common length so neither side reads past its own end.
+    auto __state_first = ::cuda::zip_transform_iterator{
+      __lex_state_fn<_Compare>{::cuda::std::move(__comp)}, ::cuda::std::move(__first1), ::cuda::std::move(__first2)};
+    auto __state_last = __state_first + __n;
+
+    // Pass 1 (early-terminating): index of the first non-equivalent pair.
+    auto __k_iter = __find_dispatch(__policy, ::cuda::std::move(__state_first), __state_last, __lex_non_zero{});
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: We will need to implement this for cuda to coalesce allocations

**Review by miscco — COMMENTED** [review]
Thanks a lot for the work. I believe we want to coalesce the two allocations we currently perform into a single one

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h (hunk 2 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: Done in `8b90fc46` — added an `__lex_ordering` enum class (`__less`/`__equal`/`__greater`). Kept the underlying -1/0/1 so the 1-element read-back recovers the sign with a plain `plus<int>`.

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h (hunk 3 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: Done. One `__temporary_storage<_OffsetType, int>` now: slot 0 = find offset, slot 1 = read-back state, shared scratch sized `max(find_bytes, reduce_bytes)`. Both passes run against the same `__get_temp_storage()`, so it is a single allocation instead of two.

File: libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h (hunk 4 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/lexicographical_compare.h]: Yep — composing the generic find_if/transform_reduce dispatches cannot share scratch, so I added a dedicated `__pstl/cuda/lexicographical_compare.h` backend and switched the frontend to select it. Serial fallback unchanged.

**Comment by edenfunf** [comment]
@miscco thanks for the review! All three points are addressed in `8b90fc46`:

- magic numbers are now an `__lex_ordering` enum class
- the CUDA path moved into a dedicated `__pstl/cuda/lexicographical_compare.h` backend that only allocates once — both passes share a single `__temporary_storage`, sized to the larger of the two

Re-ran everything locally (RTX 5070, CUDA 12.9): tests pass and memcheck is clean. Let me know if anything else needs changing.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+// Maps a pair (a, b) to its `__lex_ordering`. Used as the transform of a
+// `cuda::zip_transform_iterator`, so the per-element state is computed lazily by
+// the device kernels that walk the iterator.
+template <class _Compare>
+struct __lex_state_fn
+{
+  _Compare __comp_;
+
+  _CCCL_API explicit constexpr __lex_state_fn(_Compare __comp)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]:
```suggestion
  _CCCL_API explicit constexpr __lex_state_fn(_Compare __comp) noexcept(is_nothrow_move_constructible_v<_Compare>)
```

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    if (__comp_(__a, __b))
+    {
+      return __lex_ordering::__less;
+    }
+    if (__comp_(__b, __a))
+    {
+      return __lex_ordering::__greater;
+    }
+    return __lex_ordering::__equal;
+  }
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:112]: Nitpick: There might be pathological cases where `_Compare` has a non-const call operator

It would be safer to also have a non-const call operator

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    }
+    if (__first2 == __last2)
+    {
+      return false;
+    }
+
+    using _OffsetType = common_type_t<iter_difference_t<_InputIter1>, iter_difference_t<_InputIter2>>;
+    const auto __n1   = static_cast<_OffsetType>(::cuda::std::distance(__first1, __last1));
+    const auto __n2   = static_cast<_OffsetType>(::cuda::std::distance(__first2, __last2));
+    const auto __n    = ::cuda::std::min(__n1, __n2);
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Nitpick: We prefer more descriptive variable names like 
```suggestion
    const auto __count1   = static_cast<_OffsetType>(::cuda::std::distance(__first1, __last1));
    const auto __count2   = static_cast<_OffsetType>(::cuda::std::distance(__first2, __last2));
    const auto __count     = ::cuda::std::min(__count1, __count2);
```

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+      __stream.get());
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemcpyAsync,
+      "__pstl_cuda_lexicographical_compare: copy of find result from device to host failed",
+      ::cuda::std::addressof(__k),
+      __storage.template __get_ptr<0>(),
+      sizeof(_OffsetType),
+      ::cudaMemcpyDefault,
+      __stream.get());
+    __stream.sync();
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Suggestion: I believe we can actually optimize further here and completely drop reduce.

Rather than copying _OffsetT back to host to launch a reduce which copies it back to device we can use a custom kernel that does

```c++
_CCCL_KERNEL_ATTRIBUTES void __lex_result(...) {
    // No divergence within the common prefix: the shorter range is "less".
    if (__offset == __count)
    {
      *__offset = static_cast<_OffsetT>(__count1 < __count2);
    }
    else 
    {
      *__offset = static_cast<_OffsetT>(*(__state_first + __offset) == __lex_ordering::__less);
    }
}
```

Then we only need to copy that result back, likely saving one `cudaMemcypAsync` call

**Review by miscco — APPROVED** [review]

**Review by miscco — CHANGES_REQUESTED** [review]
Sorry approved accidentally

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 5 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Applied, thanks.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 6 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:112]: Added a non-const `operator()` overload alongside the const one.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 7 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Done, renamed to `__count1`/`__count2`/`__count`.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 8 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Dropped the reduce — a single-thread kernel now resolves the result in place from the find index, so it is just one memcpy back instead of a second launch + copy. Pushed in `9fa18810`.

**Comment by edenfunf** [comment]
No worries! Another look whenever you have time would be great. Thanks again!

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 9]
```diff
@@ -108,15 +120,20 @@ struct __lex_non_equal
...
+//   - otherwise the ordering at the divergence position decides.
+template <class _StateIter, class _OffsetType>
+_CCCL_KERNEL_ATTRIBUTES void __lexicographical_compare_result_kernel(
+  _StateIter __state_first, _OffsetType __count, bool __shorter_is_less, _OffsetType* __result)
 {
-  [[nodiscard]] _CCCL_DEVICE_API constexpr int operator()(__lex_ordering __state) const noexcept
-  {
-    return static_cast<int>(__state);
-  }
-};
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Critical: This is a kernel, so it needs to live inside the `_CCCL_BEGIN_NAMESPACE_ARCH_DEPENDENT` namespace

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+//   - if no divergence was found within the common prefix, the shorter range is
+//     "less" (`__shorter_is_less`);
+//   - otherwise the ordering at the divergence position decides.
+template <class _StateIter, class _OffsetType>
+_CCCL_KERNEL_ATTRIBUTES void __lexicographical_compare_result_kernel(
+  _StateIter __state_first, _OffsetType __count, bool __shorter_is_less, _OffsetType* __result)
+{
+  const _OffsetType __k = *__result;
+  *__result =
+    static_cast<_OffsetType>((__k == __count) ? __shorter_is_less : (*(__state_first + __k) == __lex_ordering::__less));
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Lets use slightly more descriptive names
```suggestion
  const _OffsetType __offset = *__result;
  *__result =
    static_cast<_OffsetType>((__offset == __count) ? __shorter_is_less : (*(__state_first + __offset ) == __lex_ordering::__less));
```

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 6 above)
Comment (davebayer) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:112]: Do we need the `const` one now for anything?

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    // Pass 2: resolve the answer in place on device, so only one value is copied
+    // back to the host (no second reduce launch, no offset round-trip).
+    const bool __shorter_is_less = (__count1 < __count2);
+    const void* __kernel =
+      reinterpret_cast<const void*>(&__lexicographical_compare_result_kernel<decltype(__state_first), _OffsetType>);
+    void* __kernel_args[] = {
+      const_cast<void*>(reinterpret_cast<const void*>(::cuda::std::addressof(__state_first))),
+      const_cast<void*>(reinterpret_cast<const void*>(::cuda::std::addressof(__count))),
+      const_cast<void*>(reinterpret_cast<const void*>(::cuda::std::addressof(__shorter_is_less))),
+      const_cast<void*>(reinterpret_cast<const void*>(::cuda::std::addressof(__result_ptr)))};
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: I believe we should be able to just `static_cast<void*>` and not require reinterpret_cast
```suggestion
      static_cast<void*>(::cuda::std::addressof(__state_first)),
      static_cast<void*>(::cuda::std::addressof(__count)),
      static_cast<void*>(::cuda::std::addressof(__shorter_is_less)),
      static_cast<void*>(::cuda::std::addressof(__result_ptr)));
```

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+      {
+        if (__err.status() == cudaErrorMemoryAllocation)
+        {
+          _CCCL_THROW(::std::bad_alloc);
+        }
+        else
+        {
+          throw __err;
+        }
+      }
```
Comment (davebayer) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:272]:
```suggestion
      _CCCL_TRY
      {
        return __par_impl(
          __policy,
          ::cuda::std::move(__first1),
          ::cuda::std::move(__last1),
          ::cuda::std::move(__first2),
          ::cuda::std::move(__last2),
          ::cuda::std::move(__comp));
      }
      _CCCL_CATCH (const ::cuda::cuda_error& __err)
      {
        if (__err.status() == ::cudaErrorMemoryAllocation)
        {
          _CCCL_THROW(::std::bad_alloc);
        }
        else
        {
          _CCCL_RETHROW;
        }
      }
```

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+    using _OffsetType   = common_type_t<iter_difference_t<_InputIter1>, iter_difference_t<_InputIter2>>;
+    const auto __count1 = static_cast<_OffsetType>(::cuda::std::distance(__first1, __last1));
+    const auto __count2 = static_cast<_OffsetType>(::cuda::std::distance(__first2, __last2));
+    const auto __count  = ::cuda::std::min(__count1, __count2);
+
+    // Lazy per-element ordering over the two input ranges, capped at the common
+    // length so neither side reads past its own end.
+    auto __state_first = ::cuda::zip_transform_iterator{
+      __lex_state_fn<_Compare>{::cuda::std::move(__comp)}, ::cuda::std::move(__first1), ::cuda::std::move(__first2)};
+
```
Comment (davebayer) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:171]: We need to set the context of the stream by constructing an `__ensure_current_context` object here, that will set the current device for us

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 11 above)
Comment (davebayer) [inline libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: I would just go with C-style cast in these cases
```suggestion
    const auto __kernel = (const void*)&__lexicographical_compare_result_kernel<decltype(__state_first), _OffsetType>;
    void* __kernel_args[] = {
      (void*)::cuda::std::addressof(__state_first)),
      (void*)::cuda::std::addressof(__count)),
      (void*)::cuda::std::addressof(__shorter_is_less)),
      (void*)::cuda::std::addressof(__result_ptr)));
```

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 6 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:112]: We should keep it because it lets the compiler optimize better

**Comment by bernhardmgruber** [comment]
Looking at the implementation from a bird’s-eye view, I was thinking whether we could avoid calling an extra kernel if we would extend `DeviceFind` with some kind of finalizer action when the right position is found. It turns out that's not so easy, because `DeviceFind` uses atomics in SMEM and GMEM to perform (integral) reductions when it starts to find indices where the predicate is true, so we would need to replace this by some other form of reduction that reduces index and value, since the blocks in the find kernel don't actually know with index will be the global first one (this is resolved in L2).

I think the current approach with using `DeviceFind` to find the mismatch position and then loading the two values again for the comparison is a bit wasteful, but doing better is a LOT harder. I think this is fine. We can keep a tracking issue around to revisit this.

**Comment by miscco** [comment]
yeah I believe we could do better if we extended `FindIf` but I believe  that is out of scope of this PR

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 9 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Done, moved it inside.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 10 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Done, renamed to `__offset`.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 11 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Done. Dropped the reinterpret_cast/const_cast pair and made the launched locals non-const so `static_cast<void*>` works directly.

Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h]: Went with @miscco's `static_cast<void*>` on the args. Kept `reinterpret_cast` only on the kernel function pointer, since `static_cast` can't convert a function pointer to `void*` (matches the `cudaLaunchKernel` idiom in cudax). Can switch to C-style throughout if you'd prefer.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 12 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:272]: The named `_CCCL_CATCH(...)` expands to an unbalanced brace in the no-exception (device) pass and isn't used anywhere else in the tree; the sibling PSTL cuda backends (partition/find_if/transform_reduce) all use a plain try/catch. Kept the raw try/catch for consistency but took the rest: `::cudaErrorMemoryAllocation` and a plain `throw`. Happy to revisit if you'd prefer pushing on the macro.

File: libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h (hunk 13 above)
Comment (edenfunf) [reply libcudacxx/include/cuda/std/__pstl/cuda/lexicographical_compare.h:171]: Added, right after getting the stream so the device stays current through the allocation, launches and copy-back.

**Comment by edenfunf** [comment]
@bernhardmgruber @miscco agreed, extending FindIf is out of scope here. Happy to keep a tracking issue around to revisit avoiding the second load later.

**Comment by miscco** [comment]
Sorry  I totally dropped the ball on this. Ping me if I did not respond on monday

**Comment by edenfunf** [comment]
> Sorry I totally dropped the ball on this. Ping me if I did not respond on monday

No worries at all! Hope you have a great weekend :>

**Comment by edenfunf** [comment]
@miscco When you have a chance, could you please take a look? Thanks!

**Comment by miscco** [comment]
/ok to test [0f03425](https://github.com/NVIDIA/cccl/pull/8735/commits/0f03425743658748e94f9f71b92e0d1d464496ee)

**Review by miscco — APPROVED** [review]
Looks good to me, @davebayer any comment on cuda launch here?
